<a id='toc'></a>
<center><h1>The Battle of Neighborhood</font></h1></center>
<center>Segmenting and Clustering Neighborhoods of Madrid and Barcelona</center>  


# 0. Table of Contents
- 1_    [Introduction](#introduction)
- 2_    [Objectives](#objective)
- 3_    [Requiered Libs](#libs)    
- 4_    [Data](#data)
- 4.1_  [Data. Import Credentials to IBM WATSON Notebook](#data_credentials)
- 4.2_  [Data. Import pandas dataframe](#data_pandas)
- 5_    [Methodology](#methodology)
- 6_    [Foursquare. Credentials and version](#foursquare)
- 7_    [Madrid](#madrid)
- 7.1_  [Madrid Analyze](#amadrid)
- 7.2_  [Madrid: Getting coordinates and plot the map](#cmadrid) 
- 7.3_  [Madrid: Plotting all neighbourhood into map](#nmadrid) 
- 7.4_  [Madrid: Exploring the neigborhoods](#emadrid)
- 7.5_  [Madrid: Clustering the neigborhood](#kmadrid)         
- 8_    [Barcelona](#madrid)
- 8.1_  [Barcelona Analyze](#abarcelona)
- 8.2_  [Barcelona: Getting coordinates and plot the map](#cbarcelona)
- 8.3_  [Barcelona: Plotting all neighbourhood into map](#nbarcelona)  
- 8.4_  [Barcelona: Exploring the neigborhoods](#ebarcelona)
- 8.5_  [Barcelona: Clustering the neigborhood](#kbarcelona)
- 9_    [Results](#results)
- 10_   [Discussion](#discussion)
- 11_   [Conclusion](#conclusion)

<a id='introduction'></a>
# 1. Introduction

As Spain's two premier cities, you can't go wrong whichever you choose. But if you had to select one, which should it be: Madrid or Barcelona?

Brief information about both cities:

- <b>Madrid</b> is the capital of Spain and the largest municipality in both the Community of Madrid and Spain as a whole. The city has almost 3.3 million inhabitants and a metropolitan area population of approximately 6.5 million. It is the third-largest city in the European Union (EU), smaller than only London and Berlin, and its monocentric metropolitan area is the third-largest in the EU, smaller only than those of London and Paris.The municipality covers 604.3 km2 (233.3 sq mi). The Madrid urban agglomeration has the third-largest GDP in the European Union and its influence in politics, education, entertainment, environment, media, fashion, science, culture, and the arts all contribute to its status as one of the world's major global cities. Madrid is home to two world-famous football clubs, Real Madrid and Atlético Madrid. Due to its economic output, high standard of living, and market size, Madrid is considered the leading economic hub of the Iberian Peninsula and of Southern Europe. (source: https://en.wikipedia.org/wiki/Madrid)  



- <b>Barcelona</b> is one of the popular city in Spain. It is the capital and largest city of the autonomous community of Catalonia, as well as the second most populous municipality of Spain. With a population of 1.6 million within city limits, its urban area extends to numerous neighbouring municipalities within the Province of Barcelona and is home to around 4.8 million people, making it the sixth most populous urban area in the European Union after Paris, London, Madrid, the Ruhr area and Milan.[3] It is one of the largest metropolises on the Mediterranean Sea, located on the coast between the mouths of the rivers Llobregat and Besòs, and bounded to the west by the Serra de Collserola mountain range, the tallest peak of which is 512 metres (1,680 feet) high.  (source: https://en.wikipedia.org/wiki/Barcelona)  


<a id='objective'></a>
# 2. Objective
In this project, we will study in details the area classification using Foursquare data and machine learning segmentation and clustering.
The aim of this project is to segment areas of Madrid and Barcelona based on the most common places captured from Foursquare. 

 We will try to determine, using segmentation and clustering:

- a. The similarity or dissimilarirty of both cities
- b. Classification of area located inside the city, tourism places, whether it is residential and others

<a id='libs'></a>
# 3. Requiered Libraries

In [1]:
!pip install folium
import requests                                  # library to handle requests
import pandas as pd                              # library for data analsysis
import numpy as np                               # library to handle data in a vectorized manner
import matplotlib.cm as cm
import matplotlib.colors as colors
import folium                                    # plotting library

import types
import ibm_boto3
from botocore.client import Config
from geopy import Nominatim                      # module to convert an address into latitude and longitude values
from IPython.display import Image                # libraries for displaying images
from IPython.core.display import HTML            # libraries for displaying images
from sklearn.cluster import KMeans
from pandas.io.json import json_normalize        # tranforming json file into a pandas dataframe library

 

     |████████████████████████████████| 92kB 274kB/s eta 0:00:01


<a id='data'></a>
# 4. Data

The data acquired from: 

 URLs: https://opendata-ajuntament.barcelona.cat/data/es/organization/territori?q=Barrios&sort=fecha_publicacion+desc,
       https://datos.madrid.es/sites/v/index.jsp?vgnextoid=374512b9ace9f310VgnVCM100000171f5a0aRCRD&buscar=true&Texto=Distrito&Sector=&Formato=&Periodicidad=&orderByCombo=CONTENT_INSTANCE_NAME_DECODE   
 
 The data adquired from URL, was restructured for easier manipulation and reading into files:

 - barcelona.csv 
 - madrid.csv 
     
 This files were uploaded into Github repository:
 
 https://github.com/alcupe/ds_capstone_example


<a id='data_credentials'></a>
## 4.1 Data. Import Credentials to IBM WATSON Notebook


In [2]:
# Barcelona 
# @hidden_cell
# The following code contains the credentials for a file in your IBM Cloud Object Storage.
# You might want to remove those credentials before you share your notebook.
credentials_1 = {
    'IAM_SERVICE_ID': 'iam-ServiceId-ff6aa9f5-b3fe-4e98-bbf2-af312097ec44',
    'IBM_API_KEY_ID': '6KJAVnk2cpJtasN25yOsAY8axOdBVIvfvAmsHBE0vpJB',
    'ENDPOINT': 'https://s3.eu-geo.objectstorage.service.networklayer.com',
    'IBM_AUTH_ENDPOINT': 'https://iam.eu-gb.bluemix.net/oidc/token',
    'BUCKET': 'dscoursera-donotdelete-pr-0vrzjcrfstxmux',
    'FILE': 'BARCELONA.csv'
}


In [3]:
# Madrid
# @hidden_cell
# The following code contains the credentials for a file in your IBM Cloud Object Storage.
# You might want to remove those credentials before you share your notebook.
credentials_2 = {
    'IAM_SERVICE_ID': 'iam-ServiceId-ff6aa9f5-b3fe-4e98-bbf2-af312097ec44',
    'IBM_API_KEY_ID': '6KJAVnk2cpJtasN25yOsAY8axOdBVIvfvAmsHBE0vpJB',
    'ENDPOINT': 'https://s3.eu-geo.objectstorage.service.networklayer.com',
    'IBM_AUTH_ENDPOINT': 'https://iam.eu-gb.bluemix.net/oidc/token',
    'BUCKET': 'dscoursera-donotdelete-pr-0vrzjcrfstxmux',
    'FILE': 'MADRID.csv'
}


<a id='data_pandas'></a>
## 4.2 Data. Import pandas dataframe


In [4]:
def __iter__(self): return 0

# @hidden_cell
# The following code accesses a file in your IBM Cloud Object Storage. It includes your credentials.
# You might want to remove those credentials before you share your notebook.
client_4c4528dd32ec472798b66ad7d2e3df91 = ibm_boto3.client(service_name='s3',
    ibm_api_key_id='6KJAVnk2cpJtasN25yOsAY8axOdBVIvfvAmsHBE0vpJB',
    ibm_auth_endpoint="https://iam.eu-gb.bluemix.net/oidc/token",
    config=Config(signature_version='oauth'),
    endpoint_url='https://s3.eu-geo.objectstorage.service.networklayer.com')


<a id='methodology'></a>
# 5.  Methodology

- Using the Foursquare API to explore neighborhoods in both cities, Madrid and Barcelona
- Exploring function to get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters
- K-means clustering algorithm will be use to complete this task.
- Using the Folium library to visualize the neighborhoods in Madrid and Barcelona their emerging clusters.

<a id='foursquare'></a>
# 6.  Foursquare. Credentials and version


In [5]:
CLIENT_ID = 'QL2OMYOY1B24UKBMDUGFY0HY0IKMSQIUCIPLHRK334LUQYZZ'
CLIENT_SECRET = 'K5KVFRWRKBQFTESD24TEQVG2EBWC4DCGO1HYHNS0WUOPFDTA' 
VERSION = '20180604'
LIMIT = 30

<a id='madrid'></a>
# 7.  Madrid


In [6]:
body = client_4c4528dd32ec472798b66ad7d2e3df91.get_object(Bucket='dscoursera-donotdelete-pr-0vrzjcrfstxmux',Key='MADRID.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

# MADRID dataframe
dfm= pd.read_csv(body, encoding='latin-1')
dfm.head()

City              District          Neighborhood  Postal   Latitude  \
0  Madrid  SALAMANCA             RECOLETOS              28001  40.416690   
1  Madrid             CHAMARTIN           PROSPERIDAD   28002  40.443681   
2  Madrid             CHAMARTIN         CIUDAD JARDIN   28002  40.451219   
3  Madrid             CHAMARTIN               EL VISO   28002  40.444624   
4  Madrid              CHAMBERI            RIOS ROSAS   28003  40.446613   

   Longitude  
0  -3.700346  
1  -3.673545  
2  -3.667724  
3  -3.678637  
4  -3.703114

<a id='amadrid'></a>
## 7.1 Madrid Analyze


In [7]:
print('Madrid has {} district and {} neighbohood.'.format(len(dfm['District'].unique()),dfm.shape[0]))

dfmg = dfm.groupby(["Postal", "District"])["Neighborhood"].apply(", ".join).reset_index()
dfmg.head()



Madrid has 22 district and 247 neighbohood.


Postal              District  \
0   28001  SALAMANCA              
1   28002             CHAMARTIN   
2   28003              CHAMBERI   
3   28004                CENTRO   
4   28005            ARGANZUELA   

                                     Neighborhood  
0                            RECOLETOS             
1             PROSPERIDAD, CIUDAD JARDIN, EL VISO  
2          RIOS ROSAS, VALLEHERMOSO, VALLEHERMOSO  
3                           JUSTICIA, UNIVERSIDAD  
4  IMPERIAL, ACACIAS, IMPERIAL, ACACIAS, IMPERIAL

<a id='cmadrid'></a>
## 7.2 Madrid: Getting coordinates and plot the map

In [8]:
address = "Madrid"
geolocator = Nominatim(user_agent="madrid_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('Coordinates of Madrid:  {}, {}'.format(latitude, longitude))
# create map using latitude and longitude values
madrid_m = folium.Map(location=[latitude, longitude], zoom_start=10)
madrid_m

Coordinates of Madrid:  40.4167047, -3.7035825


<a id='nmadrid'></a>
## 7.3 Madrid: Plotting all neighbourhood into map

In [9]:
for lat, lng, District, Neighborhood in zip(dfm['Latitude'], dfm['Longitude'], dfm['District'], dfm['Neighborhood']):
    label = '{}, {}'.format(Neighborhood, District)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(madrid_m)  
madrid_m

<a id='emadrid'></a>
## 7.4 Madrid: Exploring the neigborhoods


In [10]:
# Get the neighborhood's coordinates
neighborhood_n = dfm.loc[0, 'Neighborhood']
neighborhood_lat = dfm.loc[0, 'Latitude']
neighborhood_lon = dfm.loc[0, 'Longitude']



In [11]:
# Get the top 100 venues within a radius of 500 meters
# defining radius and limit of venues to get
radius=500
LIMIT=100
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, VERSION, neighborhood_lat, neighborhood_lon, radius, LIMIT)
# get the result to a json file
results = requests.get(url).json()

In [12]:
# Clean the JSON and structure it into a pandas dataframe
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
    if len(categories_list) == 0: return None
    else:
        return categories_list[0]['name']


In [13]:
venues = results['response']['groups'][0]['items']

# flatten JSON
nearby_venues = json_normalize(venues) 

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]
nearby_venues

name                  categories        lat  \
0              La Cabaña Argentina      Argentinian Restaurant  40.415696   
1                      Salmon Gurú                Cocktail Bar  40.414867   
2                            Bacoa                Burger Joint  40.416698   
3             Apple Puerta del Sol           Electronics Store  40.416804   
4                 Casino de Madrid                      Casino  40.417705   
5          La Pulpería de Victoria          Seafood Restaurant  40.416506   
6                   El Inti de Oro         Peruvian Restaurant  40.415751   
7         Arrocería Marina Ventura           Paella Restaurant  40.415353   
8               Plaza de Santa Ana                       Plaza  40.414631   
9                InClan Brutal Bar            Tapas Restaurant  40.415023   
10                    Rosi La Loca            Tapas Restaurant  40.415821   
11                      La Violeta                Dessert Shop  40.416477   
12                  Teatro Español                     Theater  40.414723   
13             Brew Wild Pizza Bar                    Beer Bar  40.414759   
14               Taberna Malaspina            Tapas Restaurant  40.415767   
15                     Askua barra                  Restaurant  40.417067   
16                  Puerta del Sol                       Plaza  40.417027   
17            El Rincon de Esteban          Spanish Restaurant  40.415803   
18                     La Venencia                    Wine Bar  40.415741   
19              Hotel Urban Madrid                       Hotel  40.416372   
20                     La Rollerie                 Coffee Shop  40.416511   
21                 Museo del Jamón            Tapas Restaurant  40.416659   
22              Pizzería El Trébol                 Pizza Place  40.416159   
23           Museo Colecciones ICO                 Art Gallery  40.416758   
24           Teatro de La Zarzuela                     Theater  40.417184   
25                 Chuka Ramen Bar                Noodle House  40.415750   
26         Restaurante Viva Madrid          Spanish Restaurant  40.415167   
27                        Serafina                  Restaurant  40.416430   
28                Lamucca de Prado                  Restaurant  40.414837   
29              Baton Rouge Madrid                Cocktail Bar  40.415826   
..                             ...                         ...        ...   
70                        llaollao          Frozen Yogurt Shop  40.417773   
71           Double Tree by Hilton                       Hotel  40.415197   
72           Teatro Reina Victoria                     Theater  40.416395   
73      Kirei Las Cortes by Kabuki         Japanese Restaurant  40.415376   
74                  La Descubierta          Spanish Restaurant  40.415198   
75         Círculo de Bellas Artes         Monument / Landmark  40.418486   
76               El Sur De Huertas                         Bar  40.413719   
77  Azotea Círculo de Bellas Artes                   Roof Deck  40.418280   
78             Mercado de la Reina            Tapas Restaurant  40.419532   
79   Club del Gourmet Corte Ingles                Gourmet Shop  40.417497   
80                   Cassette Club                   Nightclub  40.418648   
81                   Casa González               Deli / Bodega  40.413990   
82         Casa Museo Lope de Vega                      Museum  40.414326   
83                 Far home Atocha                      Hostel  40.413316   
84                           TAKOS          Mexican Restaurant  40.418938   
85                   Ciudad De Tui          Spanish Restaurant  40.418211   
86                     Cebo Madrid  Modern European Restaurant  40.416424   
87                           Tiger                   Gift Shop  40.415050   
88                      Casa Labra            Tapas Restaurant  40.417150   
89            Picalagartos Sky Bar                   Hotel Bar  40.419848   
90  Sala de Exposiciones A

In [14]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood','Neighborhood Latitude','Neighborhood Longitude', 'Venue', 'Venue Latitude', 'Venue Longitude','Venue Category']
    
    return(nearby_venues)

In [15]:
# Venues for each neighborhood

madrid_v = getNearbyVenues(names = dfm['Neighborhood'], latitudes = dfm['Latitude'], longitudes = dfm['Longitude'])

RECOLETOS           
PROSPERIDAD
CIUDAD JARDIN
EL VISO
RIOS ROSAS
VALLEHERMOSO
VALLEHERMOSO
JUSTICIA
UNIVERSIDAD
EMBAJADORES
IMPERIAL
ACACIAS
IMPERIAL
PALACIO
ACACIAS
PALACIO
EMBAJADORES
IMPERIAL
PALACIO
EL VISO
EL VISO
CASTELLANA
ESTRELLA
SANTA EUGENIA
PACIFICO
ADELFAS
CASA DE CAMPO
CASA DE CAMPO
ARGUELLES
NIÑO JESUS
IBIZA
TRAFALGAR
LUCERO
PUERTA DEL ANGEL
PUERTA DEL ANGEL
PUERTA DEL ANGEL
EMBAJADORES
CORTES
PACIFICO
ARAPILES
ARAPILES
NUEVA ESPAÑA
NUEVA ESPAÑA
NUEVA ESPAÑA
PUEBLO NUEVO
VENTAS
VENTAS
VENTAS
PALOMERAS SURESTE
PALOMERAS SURESTE
PALOMERAS SURESTE
PALOMERAS SURESTE
PALOMERAS BAJAS
PALOMERAS BAJAS
PALOMERAS BAJAS
PALOMERAS BAJAS
COMILLAS
SAN ISIDRO
SAN ISIDRO
COMILLAS
OPAÑEL
CUATRO CAMINOS
CASTILLEJOS
LOS ROSALES
BUTARQUE
SAN CRISTOBAL
SAN ANDRES
BUTARQUE
BUTARQUE
SAN ANDRES
SAN CRISTOBAL
SAN ANDRES
SAN CRISTOBAL
LOS ANGELES
SAN CRISTOBAL
SAN ANDRES
SAN ANDRES
HELLIN
HELLIN
REJAS
ARCOS
ROSAS
CANILLEJAS
ROSAS
CANILLEJAS
ROSAS
ARAVACA
ARAVACA
LAS AGUILAS
ALUCHE
CAMPAMENTO
CAM

In [16]:
madrid_v.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  RECOLETOS                          40.41669               -3.700346   
1  RECOLETOS                          40.41669               -3.700346   
2  RECOLETOS                          40.41669               -3.700346   
3  RECOLETOS                          40.41669               -3.700346   
4  RECOLETOS                          40.41669               -3.700346   

                  Venue  Venue Latitude  Venue Longitude  \
0   La Cabaña Argentina       40.415696        -3.698974   
1           Salmon Gurú       40.414867        -3.699532   
2                 Bacoa       40.416698        -3.701682   
3  Apple Puerta del Sol       40.416804        -3.702221   
4      Casino de Madrid       40.417705        -3.700037   

           Venue Category  
0  Argentinian Restaurant  
1            Cocktail Bar  
2            Burger Joint  
3       Electronics Store  
4                  Casino

In [17]:
madrid_v.shape

(5222, 7)

In [18]:
# Categories from all venues
madrid_v.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                 
ABRANTES                                 51                      51     51   
ACACIAS                                 137                     137    137   
ADELFAS                                  34                      34     34   
ALAMEDA DE OSUNA                         23                      23     23   
ALMENARA                                 63                      63     63   
ALMENDRALES                              33                      33     33   
ALUCHE                                   90                      90     90   
AMBROZ                                   11                      11     11   
APOSTOL SANTIAGO                          7                       7      7   
ARAPILES                                149                     149    149   
ARAVACA                                  30                      30     30   
ARCOS                                    35                      35     35   
ARGUELLES                                78                      78     78   
ATOCHA                                   20                      20     20   
BELLAS VISTAS                            27                      27     27   
BUENAVISTA                               21                      21     21   
BUTARQUE                                 14                      14     14   
CAMPAMENTO                               19                      19     19   
CANILLAS                                100                     100    100   
CANILLEJAS                               22                      22     22   
CASA DE CAMPO                            24                      24     24   
CASCO H.VALLECAS                         49                      49     49   
CASCO H.VICALVARO                        43                      43     43   
CASTELLANA                              100                     100    100   
CASTILLEJOS                              94                      94     94   
CIUDAD JARDIN                            30                      30     30   
COMILLAS                                 68                      68     68   
CONCEPCION                               52                      52     52   
CORTES                                  100                     100    100   
COSTILLARES                              24                      24     24   
...                                     ...                     ...    ...   
PROSPERIDAD                              78                      78     78   
PUEBLO NUEVO                             16                      16     16   
PUERTA BONITA                            24                      24     24   
PUERTA DEL ANGEL                         79                      79     79   
QUINTANA                                 41                      41     41   
RECOLETOS                               100                     100    100   
REJAS                                     6                       6      6   
RIOS ROSAS                               80                      80     80   
ROSAS                                    41                      41     41   
SAN ANDRES                               31                      31     31   
SAN CRISTOBAL                            20                      20     20   
SAN DIEGO                                 8                       8      8   
SAN FERMIN                               14                      14     14   
SAN ISIDRO                               15                      15     15   
SAN JUAN BAUTISTA                        18                      18     18   
SAN PASCUAL                              63                      63     63   
SANTA EUGENIA                            21                      21     21   
SIMANCAS                                 23                      23     23   
TIMON                                    35 

In [19]:
print('There are {} uniques categories.'.format(len(madrid_v['Venue Category'].unique())))

There are 276 uniques categories.


In [20]:
# Analyze each neighborhood
# one hot encoding
madrid_oh = pd.get_dummies(madrid_v[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
madrid_oh['Neighborhood'] = madrid_v['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [madrid_oh.columns[-1]] + list(madrid_oh.columns[:-1])
madrid_oh = madrid_oh[fixed_columns]

#examine the new dataframe size after one hot encoding
print('{} rows were returned after one hot encoding.'.format(madrid_oh.shape[0]))

#group rows by neighborhood and by taking the mean of the frequency of occurrence of each category
madrid_g = madrid_oh.groupby('Neighborhood').mean().reset_index()

#examine the new dataframe size after one hot encoding
print('{} rows were returned after grouping.'.format(madrid_g.shape[0]))

madrid_oh.head()

5222 rows were returned after one hot encoding.
103 rows were returned after grouping.


Yoga Studio  Accessories Store  African Restaurant  Airport  \
0            0                  0                   0        0   
1            0                  0                   0        0   
2            0                  0                   0        0   
3            0                  0                   0        0   
4            0                  0                   0        0   

   American Restaurant  Arcade  Arepa Restaurant  Argentinian Restaurant  \
0                    0       0                 0                       1   
1                    0       0                 0                       0   
2                    0       0                 0                       0   
3                    0       0                 0                       0   
4                    0       0                 0                       0   

   Art Gallery  Art Museum  ...  Train Station  Turkish Restaurant  \
0            0           0  ...              0                   0   
1            0           0  ...              0                   0   
2            0           0  ...              0                   0   
3            0           0  ...              0                   0   
4            0           0  ...              0                   0   

   Used Bookstore  Vegetarian / Vegan Restaurant  Venezuelan Restaurant  \
0               0                              0                      0   
1               0                              0                      0   
2               0                              0                      0   
3               0                              0                      0   
4               0                              0                      0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Wine Shop  Women's Store  
0                 0                      0         0          0              0  
1                 0                      0         0          0              0  
2                 0                      0         0          0              0  
3                 0                      0         0          0              0  
4                 0                      0         0          0              0  

[5 rows x 276 columns]

In [21]:
madrid_oh.shape

(5222, 276)

In [22]:
#print each neighborhood along with the top 5 most common venues
num_top_venues = 5

for hood in madrid_g['Neighborhood']:
    print("----"+hood+"----")
    temp = madrid_g[madrid_g['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----ABRANTES----
          venue  freq
0         Plaza  0.10
1        Bakery  0.10
2           Bar  0.08
3  Soccer Field  0.08
4  Burger Joint  0.06


----ACACIAS----
                venue  freq
0  Spanish Restaurant  0.09
1    Tapas Restaurant  0.08
2                 Bar  0.07
3          Restaurant  0.06
4         Pizza Place  0.04


----ADELFAS----
               venue  freq
0      Grocery Store  0.06
1               Café  0.06
2                Bar  0.06
3        Pizza Place  0.06
4  Food & Drink Shop  0.06


----ALAMEDA DE OSUNA----
              venue  freq
0        Restaurant  0.13
1     Metro Station  0.09
2     Grocery Store  0.09
3  Tapas Restaurant  0.09
4             Hotel  0.04


----ALMENARA----
                      venue  freq
0        Spanish Restaurant  0.14
1                Restaurant  0.08
2        Italian Restaurant  0.05
3  Mediterranean Restaurant  0.05
4                     Hotel  0.05


----ALMENDRALES----
                venue  freq
0  Spanish Restaurant  0.12
1

In [23]:
madrid_gr = madrid_oh.groupby('Neighborhood').mean().reset_index()
madrid_gr.head()

Neighborhood  Yoga Studio  Accessories Store  African Restaurant  \
0          ABRANTES          0.0                0.0                 0.0   
1           ACACIAS          0.0                0.0                 0.0   
2           ADELFAS          0.0                0.0                 0.0   
3  ALAMEDA DE OSUNA          0.0                0.0                 0.0   
4          ALMENARA          0.0                0.0                 0.0   

   Airport  American Restaurant  Arcade  Arepa Restaurant  \
0      0.0                  0.0     0.0               0.0   
1      0.0                  0.0     0.0               0.0   
2      0.0                  0.0     0.0               0.0   
3      0.0                  0.0     0.0               0.0   
4      0.0                  0.0     0.0               0.0   

   Argentinian Restaurant  Art Gallery  ...  Train Station  \
0                     0.0     0.000000  ...            0.0   
1                     0.0     0.043796  ...            0.0   
2                     0.0     0.000000  ...            0.0   
3                     0.0     0.000000  ...            0.0   
4                     0.0     0.015873  ...            0.0   

   Turkish Restaurant  Used Bookstore  Vegetarian / Vegan Restaurant  \
0                 0.0             0.0                       0.000000   
1                 0.0             0.0                       0.007299   
2                 0.0             0.0                       0.000000   
3                 0.0             0.0                       0.000000   
4                 0.0             0.0                       0.000000   

   Venezuelan Restaurant  Video Game Store  Vietnamese Restaurant  Wine Bar  \
0                    0.0               0.0               0.000000       0.0   
1                    0.0               0.0               0.000000       0.0   
2                    0.0               0.0               0.000000       0.0   
3                    0.0               0.0               0.000000       0.0   
4                    0.0               0.0               0.015873       0.0   

   Wine Shop  Women's Store  
0        0.0            0.0  
1        0.0            0.0  
2        0.0            0.0  
3        0.0            0.0  
4        0.0            0.0  

[5 rows x 276 columns]

In [24]:
def common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    return row_categories_sorted.index.values[0:num_top_venues]

In [25]:
num_top_venues = 10
indicators = ['st', 'nd', 'rd']
# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try: 
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
    neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
    neighborhoods_venues_sorted['Neighborhood'] = madrid_gr['Neighborhood']
       
for ind in np.arange(madrid_gr.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = common_venues(madrid_gr.iloc[ind, :], num_top_venues)

In [26]:
neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0          ABRANTES                 Plaza                Bakery   
1           ACACIAS    Spanish Restaurant      Tapas Restaurant   
2           ADELFAS         Grocery Store           Pizza Place   
3  ALAMEDA DE OSUNA            Restaurant         Grocery Store   
4          ALMENARA    Spanish Restaurant            Restaurant   

  3rd Most Common Venue 4th Most Common Venue     5th Most Common Venue  \
0          Soccer Field                   Bar      Colombian Restaurant   
1                   Bar            Restaurant                      Café   
2     Food & Drink Shop                   Bar                      Café   
3         Metro Station      Tapas Restaurant        Mexican Restaurant   
4    Chinese Restaurant                 Hotel  Mediterranean Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0          Burger Joint  Brazilian Restaurant         Metro Station   
1           Art Gallery         Indie Theater           Pizza Place   
2     Korean Restaurant    Chinese Restaurant        Farmers Market   
3                Bakery                  Park           Coffee Shop   
4    Italian Restaurant  Gym / Fitness Center             BBQ Joint   

  9th Most Common Venue 10th Most Common Venue  
0  Gym / Fitness Center                   Park  
1               Theater                 Market  
2           Coffee Shop              Gift Shop  
3      Asian Restaurant           Cocktail Bar  
4  Fast Food Restaurant                  Plaza

<a id='kmadrid'></a>
## 7.5 Madrid: Clustering the neigborhood

In [27]:
# Cluster neighborhoods
# Run k-means to cluster the neighborhood into 2 clusters.
# number of clusters
kclusters = 5
madrid_gc = madrid_gr.drop('Neighborhood', 1)
# run k-means 
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(madrid_gc)
# check cluster labels 
kmeans.labels_[0:10]

array([1, 1, 1, 0, 3, 3, 1, 0, 0, 3], dtype=int32)

In [28]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
neighborhoods_venues_sorted.head()

Cluster Labels      Neighborhood 1st Most Common Venue  \
0               1          ABRANTES                 Plaza   
1               1           ACACIAS    Spanish Restaurant   
2               1           ADELFAS         Grocery Store   
3               0  ALAMEDA DE OSUNA            Restaurant   
4               3          ALMENARA    Spanish Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery          Soccer Field                   Bar   
1      Tapas Restaurant                   Bar            Restaurant   
2           Pizza Place     Food & Drink Shop                   Bar   
3         Grocery Store         Metro Station      Tapas Restaurant   
4            Restaurant    Chinese Restaurant                 Hotel   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0      Colombian Restaurant          Burger Joint  Brazilian Restaurant   
1                      Café           Art Gallery         Indie Theater   
2                      Café     Korean Restaurant    Chinese Restaurant   
3        Mexican Restaurant                Bakery                  Park   
4  Mediterranean Restaurant    Italian Restaurant  Gym / Fitness Center   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0         Metro Station  Gym / Fitness Center                   Park  
1           Pizza Place               Theater                 Market  
2        Farmers Market           Coffee Shop              Gift Shop  
3           Coffee Shop      Asian Restaurant           Cocktail Bar  
4             BBQ Joint  Fast Food Restaurant                  Plaza

In [29]:
madrid_m = dfm
# merging 
madrid_m = madrid_m.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
madrid_m.head()

City              District          Neighborhood  Postal   Latitude  \
0  Madrid  SALAMANCA             RECOLETOS              28001  40.416690   
1  Madrid             CHAMARTIN           PROSPERIDAD   28002  40.443681   
2  Madrid             CHAMARTIN         CIUDAD JARDIN   28002  40.451219   
3  Madrid             CHAMARTIN               EL VISO   28002  40.444624   
4  Madrid              CHAMBERI            RIOS ROSAS   28003  40.446613   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  -3.700346               3    Spanish Restaurant                 Hotel   
1  -3.673545               1    Spanish Restaurant                   Bar   
2  -3.667724               3           Coffee Shop    Seafood Restaurant   
3  -3.678637               3    Spanish Restaurant            Restaurant   
4  -3.703114               1    Spanish Restaurant                   Bar   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0            Restaurant      Tapas Restaurant                 Plaza   
1           Supermarket            Restaurant                  Café   
2                  Café                Bakery      Asian Restaurant   
3                 Plaza           Coffee Shop                   Bar   
4      Tapas Restaurant            Restaurant         Grocery Store   

  6th Most Common Venue 7th Most Common Venue      8th Most Common Venue  \
0             Nightclub               Theater               Cocktail Bar   
1      Tapas Restaurant                   Gym              Metro Station   
2      Tapas Restaurant             Rock Club                Music Venue   
3      Theme Restaurant           Pizza Place                Snack Place   
4                  Café           Supermarket  Latin American Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0         Deli / Bodega                    Pub  
1           Coffee Shop                  Plaza  
2                   Pub        Thai Restaurant  
3          Burger Joint    Japanese Restaurant  
4           Pizza Place             Taco Place

In [30]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(
    madrid_m['Latitude'],
    madrid_m['Longitude'],
    madrid_m['Neighborhood'],
    madrid_m['Cluster Labels']):
    label = folium.Popup(str(poi) + 'Cluster' + str(cluster), parse_html=True)

    folium.CircleMarker(
                    [lat, lon],
                    radius = 5,
                    popup = label,
                    color = rainbow[cluster-1],
                    fill = True,
                    fill_color = rainbow[cluster-1],
                    fill_opacity=0.7).add_to(map_clusters)
map_clusters

In [33]:
# Madrid: Cluster 0
madrid_m.loc[madrid_m['Cluster Labels'] == 0, madrid_m.columns[[1] + list(range(5, madrid_m.shape[1]))]]

District  Longitude  Cluster Labels     1st Most Common Venue  \
48    PUENTE DE VALLECAS  -3.642605               0             Grocery Store   
49    PUENTE DE VALLECAS  -3.639028               0             Grocery Store   
50    PUENTE DE VALLECAS  -3.643950               0             Grocery Store   
51    PUENTE DE VALLECAS  -3.642006               0             Grocery Store   
57           CARABANCHEL  -3.723240               0             Grocery Store   
58           CARABANCHEL  -3.736809               0             Grocery Store   
63            VILLAVERDE  -3.686185               0             Metro Station   
65            VILLAVERDE  -3.687001               0    Furniture / Home Store   
70            VILLAVERDE  -3.685296               0    Furniture / Home Store   
72            VILLAVERDE  -3.689582               0    Furniture / Home Store   
74            VILLAVERDE  -3.684911               0    Furniture / Home Store   
77   SAN BLAS-CANILLEJAS  -3.612052               0             Grocery Store   
78             HORTALEZA  -3.612052               0             Grocery Store   
81   SAN BLAS-CANILLEJAS  -3.602577               0               Pizza Place   
82   SAN BLAS-CANILLEJAS  -3.608412               0                     Hotel   
83   SAN BLAS-CANILLEJAS  -3.608974               0               Pizza Place   
84   SAN BLAS-CANILLEJAS  -3.614890               0                     Hotel   
85   SAN BLAS-CANILLEJAS  -3.607242               0               Pizza Place   
96           CARABANCHEL  -3.742850               0             Grocery Store   
98           CARABANCHEL  -3.730927               0             Grocery Store   
100          CARABANCHEL  -3.737130               0             Grocery Store   
124            MORATALAZ  -3.643580               0             Metro Station   
129            MORATALAZ  -3.641157               0                Playground   
131            MORATALAZ  -3.637944               0             Metro Station   
132            MORATALAZ  -3.645563               0                Playground   
134    VILLA DE VALLECAS  -3.622168               0                Restaurant   
137    VILLA DE VALLECAS  -3.623809               0                Restaurant   
138    VILLA DE VALLECAS  -3.618436               0                Restaurant   
139    VILLA DE VALLECAS  -3.620773               0                Restaurant   
140    VILLA DE VALLECAS  -3.617345               0                Restaurant   
141            VICALVARO  -3.603695               0        Spanish Restaurant   
142            VICALVARO  -3.598982               0        Spanish Restaurant   
143            VICALVARO  -3.613682               0             Grocery Store   
144            VICALVARO  -3.625774               0        Spanish Restaurant   
145            VICALVARO  -3.607306               0        Spanish Restaurant   
146            VICALVARO  -3.617838               0        Spanish Restaurant   
147            VICALVARO  -3.601095               0             Grocery Store   
148            VICALVARO  -3.595314               0        Spanish Restaurant   
153            HORTALEZA  -3.659634               0        Spanish Restaurant   
159  FUENCARRAL-EL PARDO  -3.714198               0                Restaurant   
168  SAN BLAS-CANILLEJAS  -3.618281               0             Grocery Store   
172   PUENTE DE VALLECAS  -3.637380               0             Grocery Store   
173   PUENTE DE VALLECAS  -3.654489               0  Mediterranean Restaurant   
174   PUENTE DE VALLECAS  -3.658157               0  Mediterranean Restaurant   
175   PUENTE DE VALLECAS  -3.637010               0             Grocery Store   
190           VILLAVERDE  -3.693279               0             Metro Station   
191                USERA  -3.705137               0                      Park   
192           VILLAVERDE  -3.690462               0             Metro Station   
193                USERA  -3.710524            

In [32]:
# Madrid: Cluster 1
madrid_m.loc[madrid_m['Cluster Labels'] == 1, madrid_m.columns[[1] + list(range(5, madrid_m.shape[1]))]]

District  Longitude  Cluster Labels  \
0    SALAMANCA             -3.700346               1   
1               CHAMARTIN  -3.673545               1   
2               CHAMARTIN  -3.667724               1   
3               CHAMARTIN  -3.678637               1   
4                CHAMBERI  -3.703114               1   
5                CHAMBERI  -3.708835               1   
6                CHAMBERI  -3.710968               1   
7                  CENTRO  -3.699723               1   
8                  CENTRO  -3.704582               1   
9                  CENTRO  -3.705528               1   
10             ARGANZUELA  -3.714566               1   
11             ARGANZUELA  -3.703916               1   
12             ARGANZUELA  -3.720486               1   
13                 CENTRO  -3.711383               1   
14             ARGANZUELA  -3.703169               1   
15                 CENTRO  -3.712170               1   
16                 CENTRO  -3.706923               1   
17             ARGANZUELA  -3.715862               1   
18                 CENTRO  -3.713999               1   
19              CHAMARTIN  -3.681618               1   
20              CHAMARTIN  -3.686146               1   
21              SALAMANCA  -3.679669               1   
24                 RETIRO  -3.679332               1   
26        MONCLOA-ARAVACA  -3.733709               1   
27        MONCLOA-ARAVACA  -3.727309               1   
28        MONCLOA-ARAVACA  -3.714912               1   
29                 RETIRO  -3.674351               1   
31               CHAMBERI  -3.702460               1   
36                 CENTRO  -3.701275               1   
37                 CENTRO  -3.692921               1   
..                    ...        ...             ...   
65             VILLAVERDE  -3.687001               1   
70             VILLAVERDE  -3.685296               1   
72             VILLAVERDE  -3.689582               1   
74             VILLAVERDE  -3.684911               1   
79    SAN BLAS-CANILLEJAS  -3.598027               1   
88                 LATINA  -3.778899               1   
109         CIUDAD LINEAL  -3.642333               1   
110   SAN BLAS-CANILLEJAS  -3.624557               1   
115             SALAMANCA  -3.667020               1   
116   FUENCARRAL-EL PARDO  -3.702423               1   
117   FUENCARRAL-EL PARDO  -3.700653               1   
118   FUENCARRAL-EL PARDO  -3.701010               1   
119                TETUAN  -3.699342               1   
122                TETUAN  -3.693889               1   
123   FUENCARRAL-EL PARDO  -3.695908               1   
158   FUENCARRAL-EL PARDO  -3.701608               1   
187                 USERA  -3.701248               1   
196               BARAJAS  -3.584021               1   
197               BARAJAS  -3.584792               1   
198               BARAJAS  -3.591220               1   
199               BARAJAS  -3.590460               1   
206         CIUDAD LINEAL  -3.659020               1   
207                LATINA  -3.776448               1   
209                LATINA  -3.762609               1   
210                LATINA  -3.783015               1   
211                LATINA  -3.771729               1   
213            ARGANZUELA  -3.691397               1   
214            ARGANZUELA  -3.693473               1   
215            ARGANZUELA  -3.678352               1   
216            ARGANZUELA  -3.688150               1   

        1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0          Spanish Restaurant                 Hotel      Tapas Restaurant   
1          Spanish Restaurant                   Bar            Restaurant   
2                 Coffee Shop                Bakery                  Café   
3          Spanish Restaurant            Restaurant           Coffee Shop   
4          Spanish Restaurant      Tapas Restaurant                   Bar   
5                         Bar                  Café    Spanish Restaurant   
6    

In [37]:
# Madrid: Cluster 2
madrid_m.loc[madrid_m['Cluster Labels'] == 2, madrid_m.columns[[1] + list(range(5, madrid_m.shape[1]))]]

District  Longitude  Cluster Labels         1st Most Common Venue  \
32   LATINA  -3.749873               2  Theme Park Ride / Attraction   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
32            Restaurant           Supermarket      Tapas Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
32           Snack Place            Theme Park    Seafood Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
32                  Park  Fast Food Restaurant     Falafel Restaurant

In [35]:
# Madrid: Cluster 3
madrid_m.loc[madrid_m['Cluster Labels'] == 3, madrid_m.columns[[1] + list(range(5, madrid_m.shape[1]))]]

District  Longitude  Cluster Labels 1st Most Common Venue  \
0    SALAMANCA             -3.700346               3    Spanish Restaurant   
2               CHAMARTIN  -3.667724               3           Coffee Shop   
3               CHAMARTIN  -3.678637               3    Spanish Restaurant   
19              CHAMARTIN  -3.681618               3    Spanish Restaurant   
20              CHAMARTIN  -3.686146               3    Spanish Restaurant   
21              SALAMANCA  -3.679669               3    Spanish Restaurant   
28        MONCLOA-ARAVACA  -3.714912               3                 Hotel   
30                 RETIRO  -3.677873               3    Spanish Restaurant   
37                 CENTRO  -3.692921               3    Spanish Restaurant   
39               CHAMBERI  -3.709739               3    Spanish Restaurant   
40               CHAMBERI  -3.704457               3    Spanish Restaurant   
41              CHAMARTIN  -3.669068               3    Spanish Restaurant   
42              CHAMARTIN  -3.669127               3    Spanish Restaurant   
43              CHAMARTIN  -3.679542               3    Spanish Restaurant   
60            CARABANCHEL  -3.725251               3                 Plaza   
61                 TETUAN  -3.702963               3    Spanish Restaurant   
62                 TETUAN  -3.694200               3    Spanish Restaurant   
66             VILLAVERDE  -3.708957               3         Train Station   
69             VILLAVERDE  -3.705509               3         Train Station   
71             VILLAVERDE  -3.705611               3         Train Station   
75             VILLAVERDE  -3.711149               3         Train Station   
76             VILLAVERDE  -3.703196               3         Train Station   
86        MONCLOA-ARAVACA  -3.785704               3            Restaurant   
87        MONCLOA-ARAVACA  -3.771571               3            Restaurant   
102                 USERA  -3.710376               3         Grocery Store   
103                 USERA  -3.704765               3         Grocery Store   
104                 USERA  -3.699609               3    Spanish Restaurant   
105                 USERA  -3.698253               3    Spanish Restaurant   
108         CIUDAD LINEAL  -3.654404               3    Spanish Restaurant   
109         CIUDAD LINEAL  -3.642333               3    Spanish Restaurant   
..                    ...        ...             ...                   ...   
119                TETUAN  -3.699342               3    Spanish Restaurant   
122                TETUAN  -3.693889               3    Spanish Restaurant   
149             HORTALEZA  -3.634543               3           Supermarket   
150             HORTALEZA  -3.650237               3           Supermarket   
154             HORTALEZA  -3.654242               3           Supermarket   
155             HORTALEZA  -3.652228               3           Supermarket   
157   FUENCARRAL-EL PARDO  -3.696048               3    Spanish Restaurant   
160   FUENCARRAL-EL PARDO  -3.695268               3    Spanish Restaurant   
161   FUENCARRAL-EL PARDO  -3.686423               3    Spanish Restaurant   
165   SAN BLAS-CANILLEJAS  -3.628835               3            Restaurant   
169   SAN BLAS-CANILLEJAS  -3.624784               3            Restaurant   
181                TETUAN  -3.711347               3    Spanish Restaurant   
184                TETUAN  -3.703229               3    Spanish Restaurant   
187                 USERA  -3.701248               3           Coffee Shop   
189            VILLAVERDE  -3.706058               3         Train Station   
196               BARAJAS  -3.584021               3                 Hotel   
199               BARAJAS  -3.590460               3                 Hotel   
200             HORTALEZA  -3.645781               3    Spanish Restaurant   
201             HORTALEZA  -3.634109               3    Spanish Restaurant   
202             HORTALEZA  -3.647738   

In [38]:
# Madrid: Cluster 4
madrid_m.loc[madrid_m['Cluster Labels'] == 4, madrid_m.columns[[1] + list(range(5, madrid_m.shape[1]))]]

District  Longitude  Cluster Labels 1st Most Common Venue  \
64  VILLAVERDE  -3.674747               4                  Park   
67  VILLAVERDE  -3.674981               4                  Park   
68  VILLAVERDE  -3.677835               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
64        Sandwich Place                 Plaza            Sports Bar   
67        Sandwich Place                 Plaza            Sports Bar   
68        Sandwich Place                 Plaza            Sports Bar   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
64            Playground                 Trail                   Bar   
67            Playground                 Trail                   Bar   
68            Playground                 Trail                   Bar   

        8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
64  Latin American Restaurant         Grocery Store            Flea Market  
67  Latin American Restaurant         Grocery Store            Flea Market  
68  Latin American Restaurant         Grocery Store            Flea Market

<a id='barcelona'></a>
# 8.  Barcelona

In [39]:
body = client_4c4528dd32ec472798b66ad7d2e3df91.get_object(Bucket='dscoursera-donotdelete-pr-0vrzjcrfstxmux',Key='BARCELONA.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

# BARCELONA dataframe
dfb = pd.read_csv(body, encoding='latin-1')
dfb.head()

City      District                           Neighborhood  Postal  \
0  Barcelona  Ciutat Vella                               El Raval    8001   
1  Barcelona  Ciutat Vella                       El Barrio Gótico    8002   
2  Barcelona  Ciutat Vella                         La Barceloneta    8003   
3  Barcelona  Ciutat Vella  Sant Pere, Santa Caterina i la Ribera    8003   
4  Barcelona      Eixample                          El Fort Pienc    8013   

    Latitude  Longitude  
0  41.373601   2.168384  
1  41.378367   2.175353  
2  41.379057   2.190086  
3  41.386690   2.177982  
4  41.390645   2.178475

<a id='abarcelona'></a>
## 8.1 Barcelona Analyze


In [40]:
print('Barcelona has {} district and {} neighbohood.'.format(len(dfb['District'].unique()),dfb.shape[0]))

dfbg = dfb.groupby(["Postal", "District"])["Neighborhood"].apply(", ".join).reset_index()
dfbg.head()


Barcelona has 10 district and 73 neighbohood.


Postal        District                                       Neighborhood
0    8001    Ciutat Vella                                           El Raval
1    8002    Ciutat Vella                                   El Barrio Gótico
2    8003    Ciutat Vella  La Barceloneta, Sant Pere, Santa Caterina i la...
3    8004  Sants Montjuic                El Poble Sec, La Font de la Guatlla
4    8005      Sant Martí         La Vila Olímpica del Poblenou, El Poblenou

<a id='cmadrid'></a>
## 8.2 Barcelona: Getting coordinates and plot the map

In [41]:
address = "Barcelona"
geolocator = Nominatim(user_agent="barcelona_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('Coordinates of Barcelona:  {}, {}'.format(latitude, longitude))
# create map using latitude and longitude values
barcelona_m = folium.Map(location=[latitude, longitude], zoom_start=10)
barcelona_m

Coordinates of Barcelona:  41.3828939, 2.1774322


<a id='nbarcelona'></a>
## 8.3 Barcelona: Plotting all neighbourhood into map
    

In [42]:
for lat, lng, District, Neigborhood in zip(dfb['Latitude'], dfb['Longitude'], dfb['District'], dfb['Neighborhood']):
    label = '{}, {}'.format(Neigborhood, District)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(barcelona_m)  
barcelona_m

<a id='ebarcelona'></a>
## 8.4 Barcelona: Exploring the neigborhoods


In [43]:
neighborhood_n =   dfb.loc[0, 'Neighborhood']
neighborhood_lat = dfb.loc[0, 'Latitude']
neighborhood_lon = dfb.loc[0, 'Longitude']

In [44]:
# Get the top 100 venues within a radius of 500 meters
# defining radius and limit of venues to get
radius=500
LIMIT=100
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, VERSION, neighborhood_lat, neighborhood_lon, radius, LIMIT)
# get the result to a json file
results = requests.get(url).json()

In [45]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
    if len(categories_list) == 0: return None
    else:
        return categories_list[0]['name']


In [46]:
venues = results['response']['groups'][0]['items']

# flatten JSON
nearby_venues = json_normalize(venues) 

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]
nearby_venues

name                categories        lat  \
0                         Palo Cortao          Tapas Restaurant  41.372803   
1                               Koska          Tapas Restaurant  41.373124   
2                     Teatre Victoria                   Theater  41.374573   
3                          Sala Apolo              Concert Hall  41.374355   
4                     Quimet & Quimet          Tapas Restaurant  41.373997   
5                            La Chana          Tapas Restaurant  41.374469   
6                       La Confitería              Cocktail Bar  41.375520   
7                          Abirradero                   Brewery  41.374280   
8                 La Tasqueta de Blai          Tapas Restaurant  41.373358   
9                      Hotel Brummell                     Hotel  41.371698   
10        Ultramarinos Hendrick's Bar              Cocktail Bar  41.375513   
11          HelloBCN Hostel Barcelona                    Hostel  41.373592   
12                             macera              Cocktail Bar  41.375589   
13                           La Porca            Sandwich Place  41.372955   
14                             Psycho                 Rock Club  41.373072   
15                          Lascar 74       Peruvian Restaurant  41.371930   
16                             La [2]                 Nightclub  41.374466   
17                         El Pachuco        Mexican Restaurant  41.376369   
18                          La Strada               Pizza Place  41.373901   
19                Hostel One Paralelo                    Hostel  41.372265   
20          Hotel Barcelona Universal                     Hotel  41.375067   
21                              Barts                   Theater  41.375297   
22           Domino Bar by Bierzofood              Cocktail Bar  41.375852   
23            Todo Pasa por San Telmo    Argentinian Restaurant  41.373370   
24                        Bar Borrell                       Bar  41.375166   
25                          Rouge Bar                    Lounge  41.373972   
26               Taberna Blai Tonight          Tapas Restaurant  41.373635   
27                  Gran Bodega Saltó                  Wine Bar  41.372704   
28                          La Churre        Spanish Restaurant  41.372866   
29                        La Federica                   Gay Bar  41.374558   
..                                ...                       ...        ...   
61                 L'Angolino Focacce               Pizza Place  41.373331   
62                          O Meu Lar                 BBQ Joint  41.373597   
63            Andante Hotel Roof Pool                      Pool  41.376565   
64        Plaça de l'Hort de la Bomba                     Plaza  41.376916   
65            Andante Hotel Barcelona                     Hotel  41.376556   
66                       Anti-Karaoke               Karaoke Bar  41.374487   
67                            Carmesí          Tapas Restaurant  41.373769   
68      Jardins de les Tres Xemeneies                Skate Park  41.374806   
69                         Chilango’s        Mexican Restaurant  41.375776   
70                    Brew Pub Le Sec                   Brewery  41.372528   
71               Hotel Serhs Del Port                     Hotel  41.375147   
72              Terraza Hotel Miramar                 Roof Deck  41.370312   
73                             Redrum              Cocktail Bar  41.373133   
74                    Cervecería Jazz                   Brewery  41.372252   
75                          Tosta BCN                       Bar  41.373824   
76                       La Tomaquera  Mediterranean Restaurant  41.372269   
77          La Tasqueta del Paral·lel          Tapas Restaurant  41.374933   
78                  L'Atelier de Blai          Tapas Restaurant  41.374020   
79                            Malamén                  Wine Bar  41.374395   
80                           Giraluna  Mediterranean Re

In [47]:
# Explore neighborhoods
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood','Neighborhood Latitude','Neighborhood Longitude', 'Venue', 'Venue Latitude', 'Venue Longitude','Venue Category']
    
    return(nearby_venues)

In [48]:
# Venues for each neighborhood
barcelona_v = getNearbyVenues(names = dfb['Neighborhood'], latitudes = dfb['Latitude'], longitudes = dfb['Longitude'])

El Raval
El Barrio Gótico
La Barceloneta
Sant Pere, Santa Caterina i la Ribera
El Fort Pienc
La Sagrada Familia
La Dreta de l'Eixample
L'Antiga Esquerra de l'Eixample
La Nova Esquerra de l'Eixample
Sant Antoni
El Poble Sec
La Marina del Prat Vermell
La Marina de Port
La Font de la Guatlla
Hostafrancs
La Bordeta
Sants - Badal
Sants
Les Corts
La Maternitat i Sant Ramon
Pedralbes
Vallvidrera, el Tibidabo i les Planes
Sarriá
Les Tres Torres
Sant Gervasi - La Bonanova
Sant Gervasi - Galvany
El Putget i Farró
Vallcarca i els Penitents
El Coll
La Salut
La Vila de Gràcia
El Camp d'en Grassot i Gràcia Nova
El Baix Guinard
Can Baró
El Guinardó
La Font d'en Fargues
El Carmel
La Teixonera
Sant Genís dels Agudells
Montbau
La Vall d'Hebron
La Clota
Horta
Vilapicina i la Torre Llobeta
Porta
El Turó de la Peira
Can Peguera
La Guineueta
Canyelles
Les Roquetes
Verdun
La Prosperitat
La Trinitat Nova
Torre Baró
Ciutat Meridiana
Vallbona
La Trinitat Vella
Baro De Viver 
El Bon Pastor
Sant Andreu
La Sagre

In [49]:
barcelona_v.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0     El Raval              41.373601                2.168384   
1     El Raval              41.373601                2.168384   
2     El Raval              41.373601                2.168384   
3     El Raval              41.373601                2.168384   
4     El Raval              41.373601                2.168384   

             Venue  Venue Latitude  Venue Longitude    Venue Category  
0      Palo Cortao       41.372803         2.167719  Tapas Restaurant  
1            Koska       41.373124         2.166040  Tapas Restaurant  
2  Teatre Victoria       41.374573         2.168835           Theater  
3       Sala Apolo       41.374355         2.169668      Concert Hall  
4  Quimet & Quimet       41.373997         2.165522  Tapas Restaurant

In [50]:
barcelona_v.shape

(3120, 7)

In [51]:
# Categories from all venues
barcelona_v.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                           
 Provençals del Poblenou                                          77   
Baro De Viver                                                      4   
Can Baró                                                          28   
Can Peguera                                                       45   
Canyelles                                                         12   
Ciutat Meridiana                                                   6   
Diagonal Mar i el Front Marítim del Poblenou                     41   
El Baix Guinard                                                   59   
El Barrio Gótico                                                 100   
El Besòs i el Maresm                                              29   
El Bon Pastor                                                     58   
El Camp d'en Grassot i Gràcia Nova                               49   
El Camp de l'Arpa del Clot                                        58   
El Carmel                                                         12   
El Clot                                                           76   
El Coll                                                           27   
El Congrés i els Indians                                          30   
El Fort Pienc                                                    100   
El Guinardó                                                       41   
El Parc i la Llacuna del Poblenou                                 45   
El Poble Sec                                                      22   
El Poblenou                                                       73   
El Putget i Farró                                                 70   
El Raval                                                          91   
El Turó de la Peira                                               30   
Horta                                                             16   
Hostafrancs                                                      100   
L'Antiga Esquerra de l'Eixample                                  100   
La Barceloneta                                                   100   
La Bordeta                                                        24   
...                                                              ...   
La Teixonera                                                      17   
La Trinitat Nova                                                   9   
La Trinitat Vella                                                  5   
La Vall d'Hebron                                                  13   
La Verneda i la Pau                                               27   
La Vila Olímpica del Poblenou                                     43   
La Vila de Gràcia                                                 85   
Les Corts                                                         64   
Les Roquetes                                                       6   
Les Tres Torres                                                   56   
Montbau                                                            3   
Navas                                                             61   
Pedralbes                                                         59   
Porta                                                             34   
Sant Andreu                                                        7   
Sant Antoni                                                      100   
Sant Genís dels Agudells                                          4   
Sant Gervasi - Galvany                                            74   
Sant Gervasi - La Bonanova                                        42   
Sant Marta de Provençals                                          29   
Sant Pere, Santa Caterina i la Ribera                            100   
Sants                                                             72   
Sants - Badal                                                     35   
Sarriá         

In [52]:
print('There are {} uniques categories.'.format(len(barcelona_v['Venue Category'].unique())))

There are 279 uniques categories.


In [53]:
# Analyze each neighborhood

# one hot encoding
barcelona_oh = pd.get_dummies(barcelona_v[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
barcelona_oh['Neighborhood'] = barcelona_v['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [barcelona_oh.columns[-1]] + list(barcelona_oh.columns[:-1])
barcelona_oh = barcelona_oh[fixed_columns]

#examine the new dataframe size after one hot encoding
print('{} rows were returned after one hot encoding.'.format(barcelona_oh.shape[0]))

#group rows by neighborhood and by taking the mean of the frequency of occurrence of each category
barcelona_g = barcelona_oh.groupby('Neighborhood').mean().reset_index()

#examine the new dataframe size after one hot encoding
print('{} rows were returned after grouping.'.format(barcelona_g.shape[0]))


3120 rows were returned after one hot encoding.
73 rows were returned after grouping.


In [54]:
barcelona_oh.shape

(3120, 279)

In [55]:
barcelona_gr = barcelona_oh.groupby('Neighborhood').mean().reset_index()
barcelona_gr.head()

Neighborhood  Yoga Studio  Accessories Store  \
0   Provençals del Poblenou           0.0                0.0   
1             Baro De Viver           0.0                0.0   
2                   Can Baró          0.0                0.0   
3                Can Peguera          0.0                0.0   
4                  Canyelles          0.0                0.0   

   American Restaurant  Amphitheater  Animal Shelter  Arcade  \
0                  0.0           0.0             0.0     0.0   
1                  0.0           0.0             0.0     0.0   
2                  0.0           0.0             0.0     0.0   
3                  0.0           0.0             0.0     0.0   
4                  0.0           0.0             0.0     0.0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  ...  University  \
0               0.0                     0.0          0.0  ...         0.0   
1               0.0                     0.0          0.0  ...         0.0   
2               0.0                     0.0          0.0  ...         0.0   
3               0.0                     0.0          0.0  ...         0.0   
4               0.0                     0.0          0.0  ...         0.0   

   Vacation Rental  Vegetarian / Vegan Restaurant  Venezuelan Restaurant  \
0              0.0                       0.000000                    0.0   
1              0.0                       0.000000                    0.0   
2              0.0                       0.000000                    0.0   
3              0.0                       0.022222                    0.0   
4              0.0                       0.000000                    0.0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Wine Shop    Winery  \
0               0.0                    0.0       0.0        0.0  0.000000   
1               0.0                    0.0       0.0        0.0  0.000000   
2               0.0                    0.0       0.0        0.0  0.000000   
3               0.0                    0.0       0.0        0.0  0.022222   
4               0.0                    0.0       0.0        0.0  0.000000   

   Women's Store  
0       0.012987  
1       0.000000  
2       0.000000  
3       0.000000  
4       0.000000  

[5 rows x 279 columns]

In [56]:
def common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    return row_categories_sorted.index.values[0:num_top_venues]


In [57]:
num_top_venues = 10
indicators = ['st', 'nd', 'rd']
# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try: 
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
    neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
    neighborhoods_venues_sorted['Neighborhood'] = barcelona_gr['Neighborhood']
       
for ind in np.arange(barcelona_gr.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = common_venues(barcelona_gr.iloc[ind, :], num_top_venues)
    

In [58]:
neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0   Provençals del Poblenou         Clothing Store                 Hotel   
1             Baro De Viver     Spanish Restaurant         Metro Station   
2                   Can Baró    Spanish Restaurant                 Plaza   
3                Can Peguera      Tapas Restaurant    Spanish Restaurant   
4                  Canyelles                Market          Soccer Field   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Café            Restaurant           Coffee Shop   
1          Soccer Field      Tapas Restaurant           Flea Market   
2      Tapas Restaurant          Dessert Shop         Grocery Store   
3                  Café        Breakfast Spot                Bakery   
4     Food & Drink Shop                  Café         Grocery Store   

      6th Most Common Venue 7th Most Common Venue     8th Most Common Venue  \
0          Tapas Restaurant    Italian Restaurant  Mediterranean Restaurant   
1        Falafel Restaurant        Farmers Market      Fast Food Restaurant   
2        Chinese Restaurant                  Park              Soccer Field   
3               Pizza Place                   Pub              Burger Joint   
4  Mediterranean Restaurant                 Plaza                Hot Spring   

  9th Most Common Venue 10th Most Common Venue  
0                 Diner           Burger Joint  
1     Fish & Chips Shop            Fish Market  
2            Restaurant                   Café  
3        Sandwich Place            Supermarket  
4         Metro Station             Food Court

In [59]:
#print each neighborhood along with the top 5 most common venues
num_top_venues = 5

for hood in barcelona_g['Neighborhood']:
    print("----"+hood+"----")
    temp = barcelona_g[barcelona_g['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

---- Provençals del Poblenou ----
                venue  freq
0               Hotel  0.06
1      Clothing Store  0.06
2                Café  0.06
3          Restaurant  0.05
4  Italian Restaurant  0.04


----Baro De Viver ----
                venue  freq
0    Tapas Restaurant  0.25
1       Metro Station  0.25
2        Soccer Field  0.25
3  Spanish Restaurant  0.25
4         Yoga Studio  0.00


----Can Baró----
                venue  freq
0  Spanish Restaurant  0.18
1       Grocery Store  0.07
2  Chinese Restaurant  0.07
3               Plaza  0.07
4    Tapas Restaurant  0.07


----Can Peguera----
                venue  freq
0    Tapas Restaurant  0.11
1  Spanish Restaurant  0.09
2         Pizza Place  0.04
3                Café  0.04
4                 Pub  0.04


----Canyelles----
           venue  freq
0   Soccer Field  0.17
1         Market  0.17
2     Hot Spring  0.08
3  Grocery Store  0.08
4           Café  0.08


----Ciutat Meridiana----
           venue  freq
0  Metro Station  0.

<a id='kbarcelona'></a>
## 8.5 Barcelona: Clustering the neigborhood

In [60]:
# Cluster neighborhoods
# Run k-means to cluster the neighborhood into 2 clusters.
# number of clusters
kclusters = 5
barcelona_gc = barcelona_gr.drop('Neighborhood', 1)
# run k-means 
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(barcelona_gc)
# check cluster labels 
kmeans.labels_[0:10]

array([1, 4, 4, 4, 4, 0, 1, 1, 4, 1], dtype=int32)

In [61]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
neighborhoods_venues_sorted.head()

Cluster Labels               Neighborhood 1st Most Common Venue  \
0               1   Provençals del Poblenou         Clothing Store   
1               4             Baro De Viver     Spanish Restaurant   
2               4                   Can Baró    Spanish Restaurant   
3               4                Can Peguera      Tapas Restaurant   
4               4                  Canyelles                Market   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                 Hotel                  Café            Restaurant   
1         Metro Station          Soccer Field      Tapas Restaurant   
2                 Plaza      Tapas Restaurant          Dessert Shop   
3    Spanish Restaurant                  Café        Breakfast Spot   
4          Soccer Field     Food & Drink Shop                  Café   

  5th Most Common Venue     6th Most Common Venue 7th Most Common Venue  \
0           Coffee Shop          Tapas Restaurant    Italian Restaurant   
1           Flea Market        Falafel Restaurant        Farmers Market   
2         Grocery Store        Chinese Restaurant                  Park   
3                Bakery               Pizza Place                   Pub   
4         Grocery Store  Mediterranean Restaurant                 Plaza   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Mediterranean Restaurant                 Diner           Burger Joint  
1      Fast Food Restaurant     Fish & Chips Shop            Fish Market  
2              Soccer Field            Restaurant                   Café  
3              Burger Joint        Sandwich Place            Supermarket  
4                Hot Spring         Metro Station             Food Court

In [62]:
barcelona_m = dfb
# merging
barcelona_m = barcelona_m.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
barcelona_m.head()

City      District                           Neighborhood  Postal  \
0  Barcelona  Ciutat Vella                               El Raval    8001   
1  Barcelona  Ciutat Vella                       El Barrio Gótico    8002   
2  Barcelona  Ciutat Vella                         La Barceloneta    8003   
3  Barcelona  Ciutat Vella  Sant Pere, Santa Caterina i la Ribera    8003   
4  Barcelona      Eixample                          El Fort Pienc    8013   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  41.373601   2.168384               4      Tapas Restaurant   
1  41.378367   2.175353               4      Tapas Restaurant   
2  41.379057   2.190086               4      Tapas Restaurant   
3  41.386690   2.177982               1                 Hotel   
4  41.390645   2.178475               1                 Hotel   

      2nd Most Common Venue 3rd Most Common Venue  \
0                     Hotel          Cocktail Bar   
1        Spanish Restaurant    Italian Restaurant   
2  Mediterranean Restaurant     Paella Restaurant   
3          Tapas Restaurant                  Café   
4          Tapas Restaurant                Hostel   

           4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                        Brewery              Wine Bar    Mexican Restaurant   
1                            Bar                 Hotel           Art Gallery   
2                            Bar    Seafood Restaurant    Spanish Restaurant   
3  Vegetarian / Vegan Restaurant                   Bar    Spanish Restaurant   
4                            Bar           Coffee Shop    Chinese Restaurant   

           7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                        Theater                Hostel            Restaurant   
1                          Plaza           Pizza Place          Cocktail Bar   
2                     Restaurant          Burger Joint             Beach Bar   
3                          Plaza            Restaurant          Cocktail Bar   
4  Vegetarian / Vegan Restaurant                  Café    Italian Restaurant   

  10th Most Common Venue  
0     Spanish Restaurant  
1            Coffee Shop  
2            Pizza Place  
3     Italian Restaurant  
4         Breakfast Spot

In [63]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(
    barcelona_m['Latitude'],
    barcelona_m['Longitude'],
    barcelona_m['Neighborhood'],
    barcelona_m['Cluster Labels']):
    label = folium.Popup(str(poi) + 'Cluster' + str(cluster), parse_html=True)

    folium.CircleMarker(
                    [lat, lon],
                    radius = 5,
                    popup = label,
                    color = rainbow[cluster-1],
                    fill = True,
                    fill_color = rainbow[cluster-1],
                    fill_opacity=0.7).add_to(map_clusters)
map_clusters

In [67]:
# Barcelona: Cluster 0
barcelona_m.loc[barcelona_m['Cluster Labels'] == 0, barcelona_m.columns[[1] + list(range(5, barcelona_m.shape[1]))]]

District  Longitude  Cluster Labels 1st Most Common Venue  \
54  Nou Barris   2.176489               0         Metro Station   
55  Nou Barris   2.171685               0         Metro Station   
56  Nou Barris   2.183948               0                 Plaza   

   2nd Most Common Venue     3rd Most Common Venue 4th Most Common Venue  \
54                 Plaza                      Park         Train Station   
55                  Park  Mediterranean Restaurant              Building   
56         Metro Station                      Park         Train Station   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
54           Supermarket  Fast Food Restaurant           Event Space   
55         Women's Store           Flea Market        Farmers Market   
56         Women's Store     Fish & Chips Shop               Exhibit   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
54               Exhibit    Falafel Restaurant         Farmers Market  
55  Fast Food Restaurant     Fish & Chips Shop            Fish Market  
56    Falafel Restaurant        Farmers Market   Fast Food Restaurant

In [68]:
# Barcelona: Cluster 1
barcelona_m.loc[barcelona_m['Cluster Labels'] == 1, barcelona_m.columns[[1] + list(range(5, barcelona_m.shape[1]))]]

District  Longitude  Cluster Labels     1st Most Common Venue  \
3           Ciutat Vella   2.177982               1                     Hotel   
4               Eixample   2.178475               1                     Hotel   
5               Eixample   2.175593               1                      Café   
6               Eixample   2.164419               1                     Hotel   
7               Eixample   2.158034               1                     Hotel   
8               Eixample   2.153657               1                      Café   
12        Sants Montjuic   2.137948               1        Spanish Restaurant   
15        Sants Montjuic   2.140996               1                     Hotel   
16        Sants Montjuic   2.129023               1               Supermarket   
17        Sants Montjuic   2.135835               1          Tapas Restaurant   
18             Les Corts   2.131724               1                Restaurant   
19             Les Corts   2.114020               1                     Hotel   
20             Les Corts   2.123394               1                     Hotel   
21  Sarriá -Sant Gervasi   2.092727               1                 BBQ Joint   
22  Sarriá -Sant Gervasi   2.103140               1                  Mountain   
23  Sarriá -Sant Gervasi   2.128290               1        Italian Restaurant   
24  Sarriá -Sant Gervasi   2.129338               1                Restaurant   
25  Sarriá -Sant Gervasi   2.135779               1                Restaurant   
26  Sarriá -Sant Gervasi   2.143027               1                    Bakery   
28                Gracia   2.150335               1                      Park   
29                Gracia   2.148123               1                     Hotel   
31                Gracia   2.158029               1                       Bar   
32                Gracia   2.163276               1                       Bar   
42        Horta-Guinardó   2.159698               1               Coffee Shop   
58            Nou Barris   2.199296               1            Clothing Store   
59            Nou Barris   2.207082               1                      Café   
60           Sant Andreu   2.188042               1  Mediterranean Restaurant   
64           Sant Andreu   2.184065               1                    Bakery   
65           Sant Andreu   2.186908               1        Spanish Restaurant   
66            Sant Martí   2.191300               1                      Café   
67            Sant Martí   2.197048               1  Mediterranean Restaurant   
68            Sant Martí   2.197597               1        Spanish Restaurant   
69            Sant Martí   2.209500               1  Mediterranean Restaurant   
70            Sant Martí   2.216593               1            Clothing Store   
71            Sant Martí   2.199995               1  Mediterranean Restaurant   

            2nd Most Common Venue     3rd Most Common Venue  \
3                Tapas Restaurant                      Café   
4                Tapas Restaurant                    Hostel   
5              Spanish Restaurant                     Hotel   
6                Tapas Restaurant            Clothing Store   
7                Tapas Restaurant              Cocktail Bar   
8        Mediterranean Restaurant               Coffee Shop   
12                    Flea Market                      Café   
15              Food & Drink Shop               Supermarket   
16                         Bakery        Mexican Restaurant   
17                            Bar  Mediterranean Restaurant   
18             Spanish Restaurant                    Bakery   
19           Gym / Fitness Center              Tram Station   
20                     Restaurant        Spanish Restaurant   
21                  Train Station                Restaurant   
22                          Hotel                 Racetrack   
23                         Bakery       Japanese Restaurant   
24                         B

In [71]:
# Barcelona: Cluster 2
barcelona_m.loc[barcelona_m['Cluster Labels'] == 2, barcelona_m.columns[[1] + list(range(5, barcelona_m.shape[1]))]]

District  Longitude  Cluster Labels 1st Most Common Venue  \
39  Horta-Guinardó   2.130270               2        Scenic Lookout   
53      Nou Barris   2.173977               2        Scenic Lookout   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
39              Mountain         Women's Store           Flea Market   
53         Women's Store           Fish Market               Exhibit   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
39    Falafel Restaurant        Farmers Market  Fast Food Restaurant   
53    Falafel Restaurant        Farmers Market  Fast Food Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
39     Fish & Chips Shop           Fish Market            Flower Shop  
53     Fish & Chips Shop           Flea Market    Empanada Restaurant

In [72]:
# Barcelona: Cluster 3
barcelona_m.loc[barcelona_m['Cluster Labels'] == 3, barcelona_m.columns[[1] + list(range(5, barcelona_m.shape[1]))]]

District  Longitude  Cluster Labels 1st Most Common Venue  \
11  Sants Montjuic   2.159555               3                  Pier   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
11         Boat or Ferry           Art Gallery              Fountain   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
11            Food Truck            Food Stand            Food Court   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
11     Food & Drink Shop                  Food            Event Space

In [73]:
# Barcelona: Cluster 4
barcelona_m.loc[barcelona_m['Cluster Labels'] == 4, barcelona_m.columns[[1] + list(range(5, barcelona_m.shape[1]))]]

District  Longitude  Cluster Labels 1st Most Common Venue  \
0     Ciutat Vella   2.168384               4      Tapas Restaurant   
1     Ciutat Vella   2.175353               4      Tapas Restaurant   
2     Ciutat Vella   2.190086               4      Tapas Restaurant   
9         Eixample   2.164592               4      Tapas Restaurant   
10  Sants Montjuic   2.151490               4                 Plaza   
13  Sants Montjuic   2.146517               4    Spanish Restaurant   
14  Sants Montjuic   2.146076               4    Spanish Restaurant   
27          Gracia   2.133941               4                   Bar   
30          Gracia   2.154199               4                   Bar   
33          Gracia   2.159781               4    Spanish Restaurant   
34          Gracia   2.175547               4      Tapas Restaurant   
35          Gracia   2.168048               4                 Plaza   
36  Horta-Guinardó   2.151881               4                  Park   
37  Horta-Guinardó   2.146839               4                 Plaza   
38  Horta-Guinardó   2.133443               4           Supermarket   
40  Horta-Guinardó   2.144167               4    Spanish Restaurant   
41  Horta-Guinardó   2.151757               4           Coffee Shop   
43  Horta-Guinardó   2.179026               4      Tapas Restaurant   
44  Horta-Guinardó   2.173156               4                 Plaza   
45  Horta-Guinardó   2.171380               4                  Café   
46  Horta-Guinardó   2.165123               4      Tapas Restaurant   
47      Nou Barris   2.163644               4                  Park   
48      Nou Barris   2.163151               4                Market   
49      Nou Barris   2.170731               4            Skate Park   
50      Nou Barris   2.173846               4    Spanish Restaurant   
51      Nou Barris   2.179045               4    Spanish Restaurant   
52      Nou Barris   2.185450               4    Spanish Restaurant   
57      Nou Barris   2.189436               4    Spanish Restaurant   
61     Sant Andreu   2.190421               4          Burger Joint   
62     Sant Andreu   2.182797               4      Tapas Restaurant   
63     Sant Andreu   2.182967               4      Tapas Restaurant   
72      Sant Martí   2.202289               4           Coffee Shop   

       2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0                      Hotel          Cocktail Bar                    Brewery   
1         Spanish Restaurant    Italian Restaurant                        Bar   
2   Mediterranean Restaurant     Paella Restaurant                        Bar   
9                        Bar                  Café   Mediterranean Restaurant   
10      Gym / Fitness Center               Stadium                     Garden   
13                     Plaza                 Hotel                  Nightclub   
14          Tapas Restaurant           Pizza Place   Mediterranean Restaurant   
27              Cocktail Bar                  Park                      Plaza   
30          Tapas Restaurant                 Plaza                       Café   
33                     Plaza      Tapas Restaurant               Dessert Shop   
34             Grocery Store                 Plaza                 Restaurant   
35               Pizza Place        Scenic Lookout       Gym / Fitness Center   
36            Scenic Lookout                 Plaza                 Food Court   
37        Spanish Restaurant                  Park   Mediterranean Restaurant   
38              Tennis Court            Sports Bar           Tapas Restaurant   
40                       Gym                   Spa             Farmers Market   
41                    Hostel         Metro Station                 Food Court   
43                    Bakery         Grocery Store         Falafel Restaurant   
44          Tapas Restaurant                  Café              Grocery Store   
45                       Bar           Superma

<a id='results'></a>
# 9. Results

###  Madrid
    - Districts:                    22
    - Neigborhood:                  247
    - Uniques Categories:           270  
    
#### 1st. Most Common Venues
    - Cluster 0:  Grocery Store              
    - Cluster 1:  Spanish Restaurant
    - Cluster 2:  Spanish Restaurant
    - Cluster 3:  Theme Park Ride / Attraction
    - Cluster 4:  Café  
        
### Barcelona
    -Districts:                    10
    -Neigborhood:                  73
    -Uniques Categories:           278  
    
#### 1st. Most Common Venues  
    - Cluster 0:  Tapas restaurant             
    - Cluster 1:  Tapas restaurant
    - Cluster 2:  Scenic Lookout
    - Cluster 3:  Mountain
    - Cluster 4:  Park  
      
        
        
        



<a id='discussions'></a>
# 10. Discussions

- Are the results obtained sufficient to identify and distinguish different districts and describe the correlation of the most common places registered in Foursquare?
- In fact, similar cities may or may not have similar places. Then, an additional step in this classification would be to find a method to extract these common places and integrate spatial correlations between different neighborhoods or districts.
- However, the proposed segmentation and clustering is a first approach towards a quantitative and systematic comparison of the different cities.

<a id='conclusions'></a>
# 11. Conclusions

- 1. We can captured data of common places all around the world using Foursquare API
- 2. Barcelona despite having fewer districts and neighborhoods than Madrid, has a similar offer of venues to Madrid
- 3. More studies are needed to relate the acquired data and then observe them with more significant and objective results.

By Alberto Cuesta / a_cuesta@hotmail.com 